File: 백준/2352_반도체설계/Main.java
import java.io.*;
import java.util.*;

class Solution {

    private final BufferedReader reader;

    private final BufferedWriter writer;

    private int N;

    private int[] inputs;

    public Solution(BufferedReader reader) throws IOException {
        this.reader = reader;
        this.writer = new BufferedWriter(new OutputStreamWriter(System.out));
        this.N = Integer.parseInt(reader.readLine().trim());
        this.inputs = new int[N+1];
        StringTokenizer tokenizer = new StringTokenizer(reader.readLine().trim());

        for ( int i = 1 ; i <= N; i++ ) {
            inputs[i] = Integer.parseInt(tokenizer.nextToken());
        }
    }

    private int lowerBound(int[] arr, int lo, int hi, int target) {
        //int mid = lo + (hi - lo) / 2;
        int mid = 0;
        while ( lo + 1 < hi ) {
            mid = lo + (hi - lo) / 2;
            if ( arr[mid] < target ) {
                lo = mid;
            } else {
                hi = mid;
            }
        }

        return hi;
    }

    private void simulate() throws IOException {
        int[] dp = new int[N+1];
        int answer = 0;
        dp[0] = 0;
//        System.out.print("inputs : ");
//        System.out.print(Arrays.toString(inputs));
//        System.out.println();
        for ( int i = 1 ; i <= N ; ++i ) {
//            System.out.println("inputs[" + i + "] : " + inputs[i] + " max length : " + answer);
            if ( inputs[i] > dp[answer] ) {
                dp[++answer] = inputs[i];
//                System.out.println("inputs[" + i + "] is greater than dp[" + (answer-1) + "]");
            } else {
                int idx = lowerBound(dp, 0, answer, inputs[i]);
//                System.out.println("dp["+idx+"] is updated to inputs[" + i + "] : " + inputs[i]);
                dp[idx] = inputs[i];
            }
//            System.out.print("dp : ");
//            System.out.print(Arrays.toString(dp));
//            System.out.println();
        }
        writer.write(String.valueOf(answer));
        writer.write("\n");
    }

    public void run() throws IOException {
        simulate();
        writer.flush();
    }
}

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        new Solution(reader).run();
        reader.close();
    }
}

File: 백준/2352_반도체설계/main.rs
trait Solution {
    fn new() -> Self;
    fn run(&mut self);
}

struct Problem {
    n: usize,
    arr: Vec<i32>,
}

impl Problem {
    pub fn lower_bound(&self, arr: &Vec<i32>, mut lo: usize, mut hi: usize, target: i32) -> usize {
        while lo + 1 < hi {
            let mid = lo + (hi - lo) / 2;
            if arr[mid] < target {
                lo = mid;
            } else {
                hi = mid;
            }
        }
        return hi;
    }
}

impl Solution for Problem {
    fn new() -> Self {
        let n = IO::read_line().trim().parse::<usize>().unwrap();

        let arr: Vec<i32> = IO::read_line()
            .split_whitespace()
            .map(|x| (x.parse().unwrap()))
            .collect();
        Self { n: n, arr: arr }
    }

    fn run(&mut self) {
        let mut dp: Vec<i32> = vec![0; self.n + 1];
        dp[0] = 0;
        let mut answer: usize = 0;

        for i in 0..self.arr.len() {
            if dp[answer] < self.arr[i] {
                answer = answer + 1;
                dp[answer] = self.arr[i];
            } else {
                let idx = self.lower_bound(&dp, 0, answer, self.arr[i]);
                dp[idx] = self.arr[i];
            }
        }

        println!("{}", answer);
    }
}

fn main() {
    let mut solution = Problem::new();
    //solution.print_board();
    solution.run();
}

mod IO {
    pub fn read_line() -> String {
        let mut buf = String::new();
        std::io::stdin().read_line(&mut buf).unwrap();
        buf.trim().to_string()
    }

    pub fn read_copy_line(buf: &mut String) {
        std::io::stdin().read_line(buf).unwrap();
    }
}
